Troubleshooting guide for diagnosing unassigned shards (#86996)
elastic/elasticsearch: docs/reference/tab-widgets/troubleshooting/data/diagnose-unassigned-shards.asciidoc

////
[source,console]
----

> PUT my-index-000001?master_timeout=1s&timeout=1s

Author: The index will not get assigned - it's designed to remain unassigned at creation time. This shortcircuits the 30 seconds default wait time for the index to be green

> PUT my-index-000001?master_timeout=1s&timeout=1s

Reviewer: Ahh of course! Makes sense, thanks

{
  "settings": {
    "index.routing.allocation.include._name": "nonexistent_node",
    "index.routing.allocation.include._tier_preference": null
  }
}
----
////

// tag::cloud[]
In order to diagnose the unassigned shards, follow the next steps:

**Use {kib}**

//tag::kibana-api-ex[]
. Log in to the {ess-console}[{ecloud} console].
+

. On the **Elasticsearch Service** panel, click the name of your deployment. 
+

NOTE:

> If the name of your deployment is disabled your {kib} instances might be
> unhealthy, in which case please contact https://support.elastic.co[Elastic Support],
> or your deployment doesn't include {kib}, in which case all you need to do is 
> {cloud}/ec-access-kibana.html[enable Kibana first].

Reviewer 2: ```suggestion
If the name of your deployment is disabled your {kib} instances might be
unhealthy, in which case please contact https://support.elastic.co[Elastic Support]. 
If your deployment doesn't include {kib}, all you need to do is 
{cloud}/ec-access-kibana.html[enable it first].
```

. Open your deployment's side navigation menu (placed under the Elastic logo in the upper left corner)
and go to **Dev Tools > Console**.
+
[role="screenshot"]

> image::images/kibana-console.png[{kib} Console,align="center"]

Reviewer: I feel like I've seen this sentence enough that we should make it an expandable token, like we do with `{es}` or `{kib}`

Author: Would it be ok if we do this in a follow-up? (I have no idea how to do this and it'll be easy enough to, once done, replace in these guides)

Reviewer: Oh totally, not a blocker for this at all

. View the unassigned shards using the <<cat-shards,cat shards API>>.
+
[source,console]
----
GET _cat/shards?v=true&h=index,shard,prirep,state,node,unassigned.reason&s=state
----
+

> The response will look like this:

Reviewer 3: The following response is not from `GET _cat/shards` by default.  Do you mean to add `&format=json`?

+
[source,console-result]
----
{
  [
    {
      "index": "my-index-000001",
      "shard": "0",
      "prirep": "p",
      "state": "UNASSIGNED",
      "node": null,
      "unassigned.reason": "INDEX_CREATED"
    }
  ]
}
----
// TEST[skip:illustration purposes only]

+
Unassigned shards have a `state` of `UNASSIGNED`. The `prirep` value is `p` for
primary shards and `r` for replicas.
+
The index in the example has a primary shard unassigned.

. To understand why an unassigned shard is not being assigned and what action
you must take to allow {es} to assign it, use the
<<cluster-allocation-explain,cluster allocation explanation API>>.
+
[source,console]
----
GET _cluster/allocation/explain?filter_path=index,node_allocation_decisions.node_name,node_allocation_decisions.deciders.*
{
  "index": "my-index-000001", <1>
  "shard": 0, <2>
  "primary": true <3>
}
----
// TEST[skip:illustration purposes only]
+
<1> The index we want to diagnose.
+

> <2> The unassigned shard.

Reviewer 3: ```suggestion
<2> The unassigned shard ID.
```

+
<4> Whether to allocate the shard to the particular node.
+
<5> The decider which led to the `no` decision for the node.
+
<6> An explanation as to why the decider returned a `no` decision, with a helpful hint pointing to the setting that led to the decision.

. The explanation in our case indicates the index allocation configurations are not correct.
To review your allocation settings, use the <<indices-get-settings,get index
settings>> and <<cluster-get-settings,cluster get settings>> APIs.
+
[source,console]
----
GET my-index-000001/_settings?flat_settings=true&include_defaults=true

GET _cluster/settings?flat_settings=true&include_defaults=true
----
// TEST[s/^/PUT my-index-000001\n/]

. Change the settings using the <<indices-update-settings,update index
settings>> and <<cluster-update-settings,cluster update settings>> APIs to the
correct values in order to allow the index to be allocated.

For more guidance on fixing the most common causes for unassinged shards please follow
<<fix-red-yellow-cluster-status, this guide>> or contact https://support.elastic.co[Elastic Support]. 

//end::kibana-api-ex[]
// end::cloud[]

// tag::self-managed[]

> In order to get diagnoste the unassigned shards follow the next steps:

Reviewer 3: ```suggestion
In order to diagnose the unassigned shards follow the next steps:
```

Same as above

. View the unassigned shards using the <<cat-shards,cat shards API>>.
+
[source,console]
----

> GET _cat/shards?v=true&h=index,shard,prirep,state,node,unassigned.reason&s=state

Reviewer 3: ```suggestion
GET _cat/shards?v=true&h=index,shard,prirep,state,node,unassigned.reason&s=state&format=json
```

Author: I'll remove the enclosing { } to not make the request more complex

----
+
The response will look like this:
+
[source,console-result]
----
{
  [
    {
      "index": "my-index-000001",
      "shard": "0",
      "prirep": "p",
      "state": "UNASSIGNED",
      "node": null,
      "unassigned.reason": "INDEX_CREATED"
    }
  ]
}
----
// TEST[skip:illustration purposes only]

+
Unassigned shards have a `state` of `UNASSIGNED`. The `prirep` value is `p` for
primary shards and `r` for replicas.
+
The index in the example has a primary shard unassigned.

. To understand why an unassigned shard is not being assigned and what action
you must take to allow {es} to assign it, use the
<<cluster-allocation-explain,cluster allocation explanation API>>.
+
[source,console]
----

> GET _cluster/allocation/explain?filter_path=index,node_allocation_decisions.node_name,node_allocation_decisions.deciders.*

Reviewer 3: ```suggestion
GET _cluster/allocation/explain
```

{
  "index": "my-index-000001", <1>
  "shard": 0, <2>
  "primary": true <3>
}
----
// TEST[skip:illustration purposes only]
+
<1> The index we want to diagnose.
+

> <2> The unassigned shard.

Reviewer 3: ```suggestion
<2> The unassigned shard ID.
```
